File: CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using DevExpress.XtraBars;

namespace MergeMdiToolbars
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
		private DevExpress.XtraBars.BarManager barManager1;
		private DevExpress.XtraBars.BarDockControl barDockControl1;
		private DevExpress.XtraBars.BarDockControl barDockControl2;
		private DevExpress.XtraBars.BarDockControl barDockControl3;
		private DevExpress.XtraBars.BarDockControl barDockControl4;
		private DevExpress.XtraBars.BarStaticItem BarStaticItem1;
		private DevExpress.XtraBars.BarSubItem BarSubItem1;
		private DevExpress.XtraBars.BarButtonItem BarButtonItem1;
		private DevExpress.XtraBars.Bar bar1;
		private DevExpress.XtraBars.Bar bar2;
		private DevExpress.XtraBars.Bar bar3;
        private DevExpress.XtraBars.BarButtonItem barButtonItem2;
        private IContainer components;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.components = new System.ComponentModel.Container();
            this.barManager1 = new DevExpress.XtraBars.BarManager(this.components);
            this.bar1 = new DevExpress.XtraBars.Bar();
            this.BarSubItem1 = new DevExpress.XtraBars.BarSubItem();
            this.BarButtonItem1 = new DevExpress.XtraBars.BarButtonItem();
            this.bar2 = new DevExpress.XtraBars.Bar();
            this.BarStaticItem1 = new DevExpress.XtraBars.BarStaticItem();
            this.bar3 = new DevExpress.XtraBars.Bar();
            this.barButtonItem2 = new DevExpress.XtraBars.BarButtonItem();
            this.barDockControl1 = new DevExpress.XtraBars.BarDockControl();
            this.barDockControl2 = new DevExpress.XtraBars.BarDockControl();
            this.barDockControl3 = new DevExpress.XtraBars.BarDockControl();
            this.barDockControl4 = new DevExpress.XtraBars.BarDockControl();
            ((System.ComponentModel.ISupportInitialize)(this.barManager1)).BeginInit();
            this.SuspendLayout();
            // 
            // barManager1
            // 
            this.barManager1.Bars.AddRange(new DevExpress.XtraBars.Bar[] {
            this.bar1,
            this.bar2,
            this.bar3});
            this.barManager1.DockControls.Add(this.barDockControl1);
            this.barManager1.DockControls.Add(this.barDockControl2);
            this.barManager1.DockControls.Add(this.barDockControl3);
            this.barManager1.DockControls.Add(this.barDockControl4);
            this.barManager1.Form = this;
            this.barManager1.Items.AddRange(new DevExpress.XtraBars.BarItem[] {
            this.BarSubItem1,
            this.BarStaticItem1,
            this.BarButtonItem1,
            this.barButtonItem2});
            this.barManager1.MainMenu = this.bar1;
            this.barManager1.MaxItemId = 4;
            this.barManager1.StatusBar = this.bar2;
            this.barManager1.UnMerge += new DevExpress.XtraBars.BarManagerMergeEventHandler(this.barManager1_UnMerge);
            this.barManager1.Merge += new DevExpress.XtraBars.BarManagerMergeEventHandler(this.barManager1_Merge);
            // 
            // bar1
            // 
            this.bar1.BarName = "Main menu";
            this.bar1.DockCol = 0;
            this.bar1.DockRow = 0;
            this.bar1.DockStyle = DevExpress.XtraBars.BarDockStyle.Top;
            this.bar1.FloatLocation = new System.Drawing.Point(52, 165);
            this.bar1.FloatSize = new System.Drawing.Size(29, 25);
            this.bar1.LinksPersistInfo.AddRange(new DevExpress.XtraBars.LinkPersistInfo[] {
            new DevExpress.XtraBars.LinkPersistInfo(this.BarSubItem1)});
            this.bar1.OptionsBar.MultiLine = true;
            this.bar1.OptionsBar.UseWholeRow = true;
            this.bar1.Text = "Main menu";
            // 
            // BarSubItem1
            // 
            this.BarSubItem1.Caption = "Menu";
            this.BarSubItem1.Id = 1;
            this.BarSubItem1.LinksPersistInfo.AddRange(new DevExpress.XtraBars.LinkPersistInfo[] {
            new DevExpress.XtraBars.LinkPersistInfo(this.BarButtonItem1)});
            this.BarSubItem1.MergeType = DevExpress.XtraBars.BarMenuMerge.MergeItems;
            this.BarSubItem1.Name = "BarSubItem1";
            // 
            // BarButtonItem1
            // 
            this.BarButtonItem1.Caption = "Open";
            this.BarButtonItem1.Id = 2;
            this.BarButtonItem1.Name = "BarButtonItem1";
            this.BarButtonItem1.ItemClick += new DevExpress.XtraBars.ItemClickEventHandler(this.BarButtonItem1_ItemClick);
            // 
            // bar2
            // 
            this.bar2.BarName = "Status bar";
            this.bar2.CanDockStyle = DevExpress.XtraBars.BarCanDockStyle.Bottom;
            this.bar2.DockCol = 0;
            this.bar2.DockRow = 0;
            this.bar2.DockStyle = DevExpress.XtraBars.BarDockStyle.Bottom;
            this.bar2.FloatLocation = new System.Drawing.Point(65, 425);
            this.bar2.FloatSize = new System.Drawing.Size(29, 25);
            this.bar2.LinksPersistInfo.AddRange(new DevExpress.XtraBars.LinkPersistInfo[] {
            new DevExpress.XtraBars.LinkPersistInfo(this.BarStaticItem1)});
            this.bar2.OptionsBar.AllowQuickCustomization = false;
            this.bar2.OptionsBar.DrawDragBorder = false;
            this.bar2.OptionsBar.UseWholeRow = true;
            this.bar2.Text = "Status bar";
            // 
            // BarStaticItem1
            // 
            this.BarStaticItem1.Caption = "Status";
            this.BarStaticItem1.Id = 0;
            this.BarStaticItem1.Name = "BarStaticItem1";
            this.BarStaticItem1.TextAlignment = System.Drawing.StringAlignment.Near;
            // 
            // bar3
            // 
            this.bar3.BarName = "Custom 3";
            this.bar3.DockCol = 0;
            this.bar3.DockRow = 1;
            this.bar3.DockStyle = DevExpress.XtraBars.BarDockStyle.Top;
            this.bar3.FloatLocation = new System.Drawing.Point(335, 269);
            this.bar3.LinksPersistInfo.AddRange(new DevExpress.XtraBars.LinkPersistInfo[] {
            new DevExpress.XtraBars.LinkPersistInfo(this.barButtonItem2)});
            this.bar3.Text = "Custom 3";
            // 
            // barButtonItem2
            // 
            this.barButtonItem2.Caption = "Main Toolbar";
            this.barButtonItem2.Id = 3;
            this.barButtonItem2.Name = "barButtonItem2";
            // 
            // Form1
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(471, 356);
            this.Controls.Add(this.barDockControl3);
            this.Controls.Add(this.barDockControl4);
            this.Controls.Add(this.barDockControl2);
            this.Controls.Add(this.barDockControl1);
            this.IsMdiContainer = true;
            this.Name = "Form1";
            this.Text = "Form1";
            ((System.ComponentModel.ISupportInitialize)(this.barManager1)).EndInit();
            this.ResumeLayout(false);

		}
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

		private void BarButtonItem1_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{
			Form form = new Form2();
			form.MdiParent = this;
			form.Show();
		}

		private void barManager1_Merge(object sender, DevExpress.XtraBars.BarManagerMergeEventArgs e) {
			if(e.ChildManager.Form is Form2) {
				Bar newBar = new Bar(barManager1);
				newBar.BarName = "DynamicBar";
				newBar.OptionsBar.Hidden = true;
				Bar sourceBar = e.ChildManager.Bars["Custom 2"];
				
				newBar.DockStyle = BarDockStyle.Top;
				barManager1.Bars.Add(newBar);
				newBar.Merge(sourceBar);
				sourceBar.Visible = false;
			}
		}

		private void barManager1_UnMerge(object sender, DevExpress.XtraBars.BarManagerMergeEventArgs e) {
			Bar newBar = barManager1.Bars["DynamicBar"];
			if(newBar != null)
				newBar.Dispose();
		}
	}
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports DevExpress.XtraBars

Namespace MergeMdiToolbars
	''' <summary>
	''' Summary description for Form1.
	''' </summary>
	Public Class Form1
		Inherits System.Windows.Forms.Form
		Private WithEvents barManager1 As DevExpress.XtraBars.BarManager
		Private barDockControl1 As DevExpress.XtraBars.BarDockControl
		Private barDockControl2 As DevExpress.XtraBars.BarDockControl
		Private barDockControl3 As DevExpress.XtraBars.BarDockControl
		Private barDockControl4 As DevExpress.XtraBars.BarDockControl
		Private BarStaticItem1 As DevExpress.XtraBars.BarStaticItem
		Private BarSubItem1 As DevExpress.XtraBars.BarSubItem
		Private WithEvents BarButtonItem1 As DevExpress.XtraBars.BarButtonItem
		Private bar1 As DevExpress.XtraBars.Bar
		Private bar2 As DevExpress.XtraBars.Bar
		Private bar3 As DevExpress.XtraBars.Bar
		Private barButtonItem2 As DevExpress.XtraBars.BarButtonItem
		Private components As IContainer

		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()

			'
			' TODO: Add any constructor code after InitializeComponent call
			'
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Windows Form Designer generated code"
		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.components = New System.ComponentModel.Container()
			Me.barManager1 = New DevExpress.XtraBars.BarManager(Me.components)
			Me.bar1 = New DevExpress.XtraBars.Bar()
			Me.BarSubItem1 = New DevExpress.XtraBars.BarSubItem()
			Me.BarButtonItem1 = New DevExpress.XtraBars.BarButtonItem()
			Me.bar2 = New DevExpress.XtraBars.Bar()
			Me.BarStaticItem1 = New DevExpress.XtraBars.BarStaticItem()
			Me.bar3 = New DevExpress.XtraBars.Bar()
			Me.barButtonItem2 = New DevExpress.XtraBars.BarButtonItem()
			Me.barDockControl1 = New DevExpress.XtraBars.BarDockControl()
			Me.barDockControl2 = New DevExpress.XtraBars.BarDockControl()
			Me.barDockControl3 = New DevExpress.XtraBars.BarDockControl()
			Me.barDockControl4 = New DevExpress.XtraBars.BarDockControl()
			CType(Me.barManager1, System.ComponentModel.ISupportInitialize).BeginInit()
			Me.SuspendLayout()
			' 
			' barManager1
			' 
			Me.barManager1.Bars.AddRange(New DevExpress.XtraBars.Bar() { Me.bar1, Me.bar2, Me.bar3})
			Me.barManager1.DockControls.Add(Me.barDockControl1)
			Me.barManager1.DockControls.Add(Me.barDockControl2)
			Me.barManager1.DockControls.Add(Me.barDockControl3)
			Me.barManager1.DockControls.Add(Me.barDockControl4)
			Me.barManager1.Form = Me
			Me.barManager1.Items.AddRange(New DevExpress.XtraBars.BarItem() { Me.BarSubItem1, Me.BarStaticItem1, Me.BarButtonItem1, Me.barButtonItem2})
			Me.barManager1.MainMenu = Me.bar1
			Me.barManager1.MaxItemId = 4
			Me.barManager1.StatusBar = Me.bar2
'			Me.barManager1.UnMerge += New DevExpress.XtraBars.BarManagerMergeEventHandler(Me.barManager1_UnMerge);
'			Me.barManager1.Merge += New DevExpress.XtraBars.BarManagerMergeEventHandler(Me.barManager1_Merge);
			' 
			' bar1
			' 
			Me.bar1.BarName = "Main menu"
			Me.bar1.DockCol = 0
			Me.bar1.DockRow = 0
			Me.bar1.DockStyle = DevExpress.XtraBars.BarDockStyle.Top
			Me.bar1.FloatLocation = New System.Drawing.Point(52, 165)
			Me.bar1.FloatSize = New System.Drawing.Size(29, 25)
			Me.bar1.LinksPersistInfo.AddRange(New DevExpress.XtraBars.LinkPersistInfo() { New DevExpress.XtraBars.LinkPersistInfo(Me.BarSubItem1)})
			Me.bar1.OptionsBar.MultiLine = True
			Me.bar1.OptionsBar.UseWholeRow = True
			Me.bar1.Text = "Main menu"
			' 
			' BarSubItem1
			' 
			Me.BarSubItem1.Caption = "Menu"
			Me.BarSubItem1.Id = 1
			Me.BarSubItem1.LinksPersistInfo.AddRange(New DevExpress.XtraBars.LinkPersistInfo() { New DevExpress.XtraBars.LinkPersistInfo(Me.BarButtonItem1)})
			Me.BarSubItem1.MergeType = DevExpress.XtraBars.BarMenuMerge.MergeItems
			Me.BarSubItem1.Name = "BarSubItem1"
			' 
			' BarButtonItem1
			' 
			Me.BarButtonItem1.Caption = "Open"
			Me.BarButtonItem1.Id = 2
			Me.BarButtonItem1.Name = "BarButtonItem1"
'			Me.BarButtonItem1.ItemClick += New DevExpress.XtraBars.ItemClickEventHandler(Me.BarButtonItem1_ItemClick);
			' 
			' bar2
			' 
			Me.bar2.BarName = "Status bar"
			Me.bar2.CanDockStyle = DevExpress.XtraBars.BarCanDockStyle.Bottom
			Me.bar2.DockCol = 0
			Me.bar2.DockRow = 0
			Me.bar2.DockStyle = DevExpress.XtraBars.BarDockStyle.Bottom
			Me.bar2.FloatLocation = New System.Drawing.Point(65, 425)
			Me.bar2.FloatSize = New System.Drawing.Size(29, 25)
			Me.bar2.LinksPersistInfo.AddRange(New DevExpress.XtraBars.LinkPersistInfo() { New DevExpress.XtraBars.LinkPersistInfo(Me.BarStaticItem1)})
			Me.bar2.OptionsBar.AllowQuickCustomization = False
			Me.bar2.OptionsBar.DrawDragBorder = False
			Me.bar2.OptionsBar.UseWholeRow = True
			Me.bar2.Text = "Status bar"
			' 
			' BarStaticItem1
			' 
			Me.BarStaticItem1.Caption = "Status"
			Me.BarStaticItem1.Id = 0
			Me.BarStaticItem1.Name = "BarStaticItem1"
			Me.BarStaticItem1.TextAlignment = System.Drawing.StringAlignment.Near
			' 
			' bar3
			' 
			Me.bar3.BarName = "Custom 3"
			Me.bar3.DockCol = 0
			Me.bar3.DockRow = 1
			Me.bar3.DockStyle = DevExpress.XtraBars.BarDockStyle.Top
			Me.bar3.FloatLocation = New System.Drawing.Point(335, 269)
			Me.bar3.LinksPersistInfo.AddRange(New DevExpress.XtraBars.LinkPersistInfo() { New DevExpress.XtraBars.LinkPersistInfo(Me.barButtonItem2)})
			Me.bar3.Text = "Custom 3"
			' 
			' barButtonItem2
			' 
			Me.barButtonItem2.Caption = "Main Toolbar"
			Me.barButtonItem2.Id = 3
			Me.barButtonItem2.Name = "barButtonItem2"
			' 
			' Form1
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(471, 356)
			Me.Controls.Add(Me.barDockControl3)
			Me.Controls.Add(Me.barDockControl4)
			Me.Controls.Add(Me.barDockControl2)
			Me.Controls.Add(Me.barDockControl1)
			Me.IsMdiContainer = True
			Me.Name = "Form1"
			Me.Text = "Form1"
			CType(Me.barManager1, System.ComponentModel.ISupportInitialize).EndInit()
			Me.ResumeLayout(False)

		End Sub
		#End Region

		''' <summary>
		''' The main entry point for the application.
		''' </summary>
		<STAThread> _
		Shared Sub Main()
			Application.Run(New Form1())
		End Sub

		Private Sub BarButtonItem1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles BarButtonItem1.ItemClick
			Dim form As Form = New Form2()
			form.MdiParent = Me
			form.Show()
		End Sub

		Private Sub barManager1_Merge(ByVal sender As Object, ByVal e As DevExpress.XtraBars.BarManagerMergeEventArgs) Handles barManager1.Merge
			If TypeOf e.ChildManager.Form Is Form2 Then
				Dim newBar As New Bar(barManager1)
				newBar.BarName = "DynamicBar"
				newBar.OptionsBar.Hidden = True
				Dim sourceBar As Bar = e.ChildManager.Bars("Custom 2")

				newBar.DockStyle = BarDockStyle.Top
				barManager1.Bars.Add(newBar)
				newBar.Merge(sourceBar)
				sourceBar.Visible = False
			End If
		End Sub

		Private Sub barManager1_UnMerge(ByVal sender As Object, ByVal e As DevExpress.XtraBars.BarManagerMergeEventArgs) Handles barManager1.UnMerge
			Dim newBar As Bar = barManager1.Bars("DynamicBar")
			If newBar IsNot Nothing Then
				newBar.Dispose()
			End If
		End Sub
	End Class
End Namespace
